Implement JMS source/sink plugin

dardanxh opened the pull request:
There are a couple of points still open for discussion. These points are mentioned in the last section of the [engineering doc](https://docs.google.com/document/d/1ZpXn_tWk71eVJ_A2x5zUZGP5thA7S7uV/edit#heading=h.gjdgxs):

Jira: https://cdap.atlassian.net/browse/PLUGIN-597

dardanxh committed:
Implement JMS source/sink plugin

CuriousVini commented on checkstyle.xml:
```diff
@@ -0,0 +1,406 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  Copyright © 2014-2015 Cask Data, Inc.
```
2020

dardanxh replied:
Should this be 2021 now.

CuriousVini replied:
right.

CuriousVini commented on docs/JMS-batchsink.md:
```diff
@@ -0,0 +1,57 @@
+# JMS Streaming Source
```
batch sink?

dardanxh replied:
Fixed

CuriousVini commented on docs/JMS-batchsink.md:
```diff
@@ -0,0 +1,57 @@
+
+Properties
+----------
+**Connection Factory**: Name of the connection factory. If not specified *ConnectionFactory* if considered by default.
```
`If not specified *ConnectionFactory* if considered by default.` - what does this mean?

dardanxh replied:
Fixed. The second *if* (typo)  makes this sound confusing. 

It means: If not specified, the value *ConnectionFactory* is considered by default.

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,426 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
missing pom license

dardanxh replied:
Added

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,426 @@
+        <version>1.1.0</version>
+        <configuration>
+          <cdapArtifacts>
+            <parent>system:cdap-data-pipeline[6.1.0-SNAPSHOT,7.0.0-SNAPSHOT)</parent>
```
lower bound should be same as cdap version being used.

dardanxh replied:
Changed lower bound to 6.4.0-SNAPSHOT

CuriousVini commented on src/main/java/io/cdap/plugin/jms/common/JMSConfig.java:
```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright © 2016-2020 Cask Data, Inc.
```
2020

dardanxh replied:
This is already 2020

CuriousVini commented on src/test/java/JMSStreamingSourceTest.java:
```diff
@@ -0,0 +1,69 @@
+import org.junit.Assert;
+import org.junit.Test;
+import org.mockito.internal.util.reflection.FieldSetter;
+
```
need java doc

dardanxh replied:
Added

CuriousVini commented on suppressions.xml:
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0"?>
+
+<!--
+  Copyright © 2014-2015 Cask Data, Inc.
```
2020

dardanxh replied:
Fixed

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,426 @@
+  <dependencies>
+    <dependency>
+      <groupId>org.apache.logging.log4j</groupId>
+      <artifactId>log4j-api</artifactId>
```
why are logging dependencies needed as compile time dependencies?

dardanxh replied:
Removed explicit log4j dependencies from pom. They are already included by other libs.

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,426 @@
+        </exclusion>
+        <!--
+          Need to exclude this since Spark brings in an older version that makes
+          S3 not work in standalone. This doesn't affect distributed mode.
```
Is this copy/pasted from cdap pom?

dardanxh replied:
Nope. Adapted it mainly from splunk plugin. Already cleaned the pom now.

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,426 @@
+    <dependency>
+      <groupId>io.cdap.cdap</groupId>
+      <artifactId>cdap-etl-api-spark</artifactId>
+      <version>6.4.0-SNAPSHOT</version>
```
should use cdap version property

dardanxh replied:
Fixed

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,426 @@
+    <dependency>
+      <groupId>io.cdap.cdap</groupId>
+      <artifactId>cdap-api-common</artifactId>
+      <version>6.4.0-SNAPSHOT</version>
```
should use cdap version property

dardanxh replied:
Fixed

CuriousVini commented on docs/JMS-batchsink.md:
```diff
@@ -0,0 +1,57 @@
+
+**Type**: Queue or Topic. Queue is considered by default.
+
+**Destination**: Queue/Topic name.
```
Is the topic created if its not preset?

dardanxh replied:
No, topic is not created if not present. This isn't specified in the PRD. Should I implement this @CuriousVini?

CuriousVini replied:
Yes, we should create the topic/queue if not present. Also we should document that behavior

CuriousVini replied:
Missed?

CuriousVini commented on src/main/java/io/cdap/plugin/jms/common/JMSConfig.java:
```diff
@@ -0,0 +1,250 @@
+
+  @Name(NAME_TYPE)
+  @Description("Queue or Topic.")
+  private String type;
```
Does this need to be required? Since we already have a default for it?

CuriousVini replied:
Can this property not have macro?

dardanxh replied:
No this doesn't need to be required. Added `@Nullable`.
Yes this property should have the `@Macro` annotation specified.

CuriousVini commented on src/main/java/io/cdap/plugin/jms/common/JMSConfig.java:
```diff
@@ -0,0 +1,250 @@
+
+  @Name(NAME_MESSAGE_TYPE)
+  @Description("Supports the following message types: Message, Text, Bytes, Map.")
+  private String messageType;
```
should this be macro and optional?

dardanxh replied:
Yes, you are right. Added `@Macro` and `@Nullable`.

CuriousVini commented on src/main/java/io/cdap/plugin/jms/common/JMSConfig.java:
```diff
@@ -0,0 +1,250 @@
+  @Name(NAME_SCHEMA)
+  @Nullable
+  @Description("Specifies the schema of the records outputted from this plugin.")
+  private String schema;
```
why cant schema be macro?

dardanxh replied:
At schema, only the `payload` field is expected to take different data types. `payload` data type is automatically changed corresponding to the `messageType` the user selects. Since the `messageType` is already a macro, I didn't specify macro at schema.

Should we allow the users to manually customise the output schema (add new fields, manually change data types,...). This isn't specified in the PRD.

CuriousVini replied:
could you please followup with @bdmogal on the product requirements? I think this should be allowed to be a macro.

CuriousVini commented on src/main/java/io/cdap/plugin/jms/common/JMSConfig.java:
```diff
@@ -0,0 +1,250 @@
+    return jndiContextFactory;
+  }
+
+  public String getJndiUsername() {
```
should have`@Nullable` annotation for all the fields that can be null

CuriousVini replied:
Also please make sure callers are handling null values without throwing NPE

dardanxh replied:
Done.

CuriousVini reviewed (changes requested):
1. Please add license to all the code files with year 2020
2. Please only add needed dependencies in the pom
3. What is the number of records streaming source has been tested with? Same for batch sink.

dardanxh commented:
> 3. What is the number of records streaming source has been tested with? Same for batch sink.

@CuriousVini I've tested with 1M records. Should I test with much more (10M, 100M)?

chtyim commented on src/main/java/io/cdap/plugin/jms/common/JMSConnection.java:
```diff
@@ -0,0 +1,177 @@
+        try {
+            return new InitialContext(properties);
+        } catch (NamingException e) {
+            LOG.error("Initial context couldn't get created.", e);
```
Don't log and throw exception

chtyim commented on src/main/java/io/cdap/plugin/jms/common/JMSConnection.java:
```diff
@@ -0,0 +1,177 @@
+            return new InitialContext(properties);
+        } catch (NamingException e) {
+            LOG.error("Initial context couldn't get created.", e);
+            throw new RuntimeException(e.getMessage());
```
The cause should be passed together with a message that provide more context. E.g.

```
throw new RuntimeException("Failed to create initial context for provider URL " + config.getProviderUrl() + " with principal " + config.getJndiUsername(), e);
```

dardanxh replied:
Done

chtyim commented on src/main/java/io/cdap/plugin/jms/sink/JMSOutputFormat.java:
```diff
@@ -0,0 +1,71 @@
+ */
+public class JMSOutputFormat extends OutputFormat<NullWritable, Text> {
+
+    @Override
```
Wrong indentation. It should be 2 spaces.

chtyim commented on src/main/java/io/cdap/plugin/jms/source/JMSReceiver.java:
```diff
@@ -0,0 +1,90 @@
+  public void onStart() {
+
+    this.jmsConnection = new JMSConnection(config);
+
```
Remove unnecessary empty lines

dardanxh replied:
Removed

chtyim commented on src/main/java/io/cdap/plugin/jms/source/JMSSourceUtils.java:
```diff
@@ -0,0 +1,142 @@
+    IllegalArgumentException {
+    if (message instanceof BytesMessage && config.getMessageType().equals(JMSMessageType.BYTES.getName())) {
+      return convertByteMessage(message, config);
+    } else if (message instanceof MapMessage && config.getMessageType().equals(JMSMessageType.MAP.getName())) {
```
No need to have `else` after an `if` block that returns. It is cleaner to have:

```
if (...) {
  return ...;
}
if (...) {
  return ...;
}
```

dardanxh replied:
Agree. Fixed this.

CuriousVini commented on src/main/java/io/cdap/plugin/jms/source/ReferenceStreamingSource.java:
```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright © 2016-2020 Cask Data, Inc.
```
all the license should have `Copyright © 2021 Cask Data, Inc.` as mentioned in the previous comments

CuriousVini commented on docs/JMS-streamingsource.md:
```diff
@@ -0,0 +1,102 @@
+
+**Type**: Queue or Topic. Queue is considered by default.
+
+**Destination**: Queue/Topic name.
```
Destination? What does that mean in terms of streaming source?

CuriousVini replied:
This property is not part of PRD. Has this been agreed upon by product team?

dardanxh replied:
1. **Destination** is not an appropriate name here.  I'll change this. Since I've got only one config class defined for both source and sink, I'll rename this to **Name** and indicate in the description what this means.

2. This property is added by me. I've written a couple of open issues in the engineering doc attached in this PR's description and this is one of them. I've emailed this issue to Bhooshan and Bhooshan CC-ed Chai Pydimukkala on that email. These issues aren't addressed yet.

Since you want to read from a topic/queue, you should provide its name to the plugin in order to do so. That's why I suggested adding this new param.

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,349 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  Copyright © 2015-2020 Cask Data, Inc.
```
2021?

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,349 @@
+      <groupId>io.cdap.cdap</groupId>
+      <artifactId>cdap-data-streams</artifactId>
+      <version>${cdap.version}</version>
+      <scope>test</scope>
```
why is this added in test scope?

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,349 @@
+      <groupId>com.google.guava</groupId>
+      <artifactId>guava</artifactId>
+      <version>13.0.1</version>
+      <scope>compile</scope>
```
why older version of guava is this needed in compile scope?

dardanxh replied:
Upgraded to 27.0.1-jre. Removed the compile scope.

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,349 @@
+      <groupId>io.cdap.cdap</groupId>
+      <artifactId>cdap-etl-api-spark</artifactId>
+      <version>${cdap.version}</version>
+      <scope>compile</scope>
```
there is no need to provide this explicitly. default is compile

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,349 @@
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <version>${junit.version}</version>
```
why is this in compile scope? It should be in test scope

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,349 @@
+    <widgets.dir>widgets</widgets.dir>
+    <docs.dir>docs</docs.dir>
+    <main.basedir>${project.basedir}</main.basedir>
+    <cdap.version>6.4.0-SNAPSHOT</cdap.version>
```
this does not match version range in parent artifacts. Also should not use snapshot versions

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,349 @@
+    <docs.dir>docs</docs.dir>
+    <main.basedir>${project.basedir}</main.basedir>
+    <cdap.version>6.4.0-SNAPSHOT</cdap.version>
+    <cdap.plugin.version>2.6.0-SNAPSHOT</cdap.plugin.version>
```
should not use snapshot versions.

dardanxh replied:
@CuriousVini which version should I use for `cdap.plugin.version`?

CuriousVini replied:
Missed?

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,349 @@
+    <main.basedir>${project.basedir}</main.basedir>
+    <cdap.version>6.4.0-SNAPSHOT</cdap.version>
+    <cdap.plugin.version>2.6.0-SNAPSHOT</cdap.plugin.version>
+    <spark.version>1.6.1</spark.version>
```
why are we using spark 1.6?

dardanxh replied:
I'll bump to 2.3.1. Any objections with this version?

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,349 @@
+    <cdap.version>6.4.0-SNAPSHOT</cdap.version>
+    <cdap.plugin.version>2.6.0-SNAPSHOT</cdap.plugin.version>
+    <spark.version>1.6.1</spark.version>
+    <hadoop.version>2.3.0</hadoop.version>
```
why are we using hadoop 2.3?

dardanxh replied:
I'll bump to 2.9.2. Any objections with this version?

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,349 @@
+    <dependency>
+      <groupId>io.cdap.http</groupId>
+      <artifactId>netty-http</artifactId>
+      <version>${netty-http.version}</version>
```
why is this needed in compile scope?

dardanxh replied:
I'll bump to 2.3.1. Any objections with this version?

dardanxh replied:
Switched to test.

CuriousVini replied:
why is this needed in test?

CuriousVini reviewed (changes requested):
1. Please make sure license year is correct - 2021
2. Please clean up pom file with only required dependencies
3. What is the CDF version for which this plugin needs to be available? 6.1 onwards?
4. Should not use snapshot versions in pom and pipeline json

CuriousVini commented:
Also what is the scale at which performance tests were performed for:
- JMS Batch sink
- JMS Streaming source

CuriousVini commented:
Also please make sure plugin icons are included

CuriousVini commented:
> Also what is the scale at which performance tests were performed for:
> 
> * JMS Batch sink
> * JMS Streaming source

could you please provide information on this?

CuriousVini commented:
https://cdap.atlassian.net/browse/PLUGIN-550 - please add the PR to this jira

dardanxh commented:
> > Also what is the scale at which performance tests were performed for:
> > 
> > * JMS Batch sink
> > * JMS Streaming source

@CuriousVini 

The recent test performed is done locally. An ActiveMQ broker is started, and a message producer generates 1M messages to a topic where the JMS Source Plugin is subscribed. The JMS Sink Plugin then sinks the same messages to another topic. 

Testing in CDF is not yet done. As I'm been told from the team, there are firewall issues in CDF blocking this. 

Other than this, testing the JMS Source/Sink Plugins in CDF would require having a JMS broker (eg. ActiveMQ) up and running. I am now investigating how to make this work out.

CuriousVini commented:
> > > Also what is the scale at which performance tests were performed for:
> > > 
> > > * JMS Batch sink
> > > * JMS Streaming source
> 
> @CuriousVini
> 
> The recent test performed is done locally. An ActiveMQ broker is started, and a message producer generates 1M messages to a topic where the JMS Source Plugin is subscribed. The JMS Sink Plugin then sinks the same messages to another topic.
> 
> Testing in CDF is not yet done. As I'm been told from the team, there are firewall issues in CDF blocking this.
> 
> Other than this, testing the JMS Source/Sink Plugins in CDF would require having a JMS broker (eg. ActiveMQ) up and running. I am now investigating how to make this work out.

CDF 6.2+ versions are not affected by firewall rules. We should perform performance tests for both the plugins on cloud environment.

CuriousVini commented:
> https://cdap.atlassian.net/browse/PLUGIN-550 - please add the PR to this jira

Please add this PR to the jira and jira in the PR description.

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,350 @@
+    <widgets.dir>widgets</widgets.dir>
+    <docs.dir>docs</docs.dir>
+    <main.basedir>${project.basedir}</main.basedir>
+    <cdap.version>6.2.3</cdap.version>
```
this should be 6.1.2

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,350 @@
+    </dependency>
+    <dependency>
+      <groupId>io.cdap.cdap</groupId>
+      <artifactId>cdap-data-streams</artifactId>
```
why is this dependency needed?

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,350 @@
+    <dependency>
+      <groupId>org.apache.activemq</groupId>
+      <artifactId>activemq-all</artifactId>
+      <version>5.11.1</version>
```
this should be a property

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,350 @@
+    </dependency>
+    <dependency>
+      <groupId>io.cdap.cdap</groupId>
+      <artifactId>cdap-etl-api-spark</artifactId>
```
this should be provided

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,350 @@
+      <version>${junit.version}</version>
+      <scope>test</scope>
+    </dependency>
+        <dependency>
```
indentation

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,350 @@
+        <version>1.1.0</version>
+        <configuration>
+          <cdapArtifacts>
+            <parent>system:cdap-data-pipeline[6.2.3, 7.0.0)</parent>
```
should be 6.1.2

CuriousVini commented on pom.xml:
```diff
@@ -0,0 +1,350 @@
+        <configuration>
+          <cdapArtifacts>
+            <parent>system:cdap-data-pipeline[6.2.3, 7.0.0)</parent>
+            <parent>system:cdap-data-streams[6.2.3, 7.0.0)</parent>
```
same here

CuriousVini commented on src/main/java/io/cdap/plugin/jms/sink/JMSBatchSink.java:
```diff
@@ -0,0 +1,85 @@
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    config.validate(pipelineConfigurer.getStageConfigurer().getFailureCollector());
```
create a new variable for FailureCollector

CuriousVini commented on src/main/java/io/cdap/plugin/jms/common/JMSConfig.java:
```diff
@@ -0,0 +1,255 @@
+
+  @Name(NAME_TYPE)
+  @Description("Queue or Topic.")
+  @Nullable
```
type is a required field in PRD

CuriousVini commented on src/main/java/io/cdap/plugin/jms/common/JMSConfig.java:
```diff
@@ -0,0 +1,255 @@
+
+  @Name(NAME_MESSAGE_TYPE)
+  @Description("Supports the following message types: Message, Text, Bytes, Map.")
+  @Nullable
```
message type is a required field

CuriousVini reviewed (changes requested):
1. Why are hadoop dependencies needed? 
2. Please make sure all the fields in source and sink comply with PRD requirements
3. Why schema is a text box?
4. Reference name should be first property

Please make sure this is reviewed internally

![image](https://user-images.githubusercontent.com/14131070/105226093-619bb500-5b14-11eb-9554-fabfd48eba96.png)

dardanxh reviewed (commented):
> 1. Why are hadoop dependencies needed?
> 2. Please make sure all the fields in source and sink comply with PRD requirements
> 3. Why schema is a text box?
> 4. Reference name should be first property
> 
> Please make sure this is reviewed internally
> 
> ![image](https://user-images.githubusercontent.com/14131070/105226093-619bb500-5b14-11eb-9554-fabfd48eba96.png)

@CuriousVini 
1. As a way to use newer versions of hadoop rather than the one provided by default. POM is cleaned according to the comment you've made. 

2. That is made sure. Furthermore there are still some points opened that are to be decided (please see below). Need comments on these.

> Issue 1: A new property called Name is introduced. This is used to specify the Topic or Queue name you want to produce/consume messages to/from. This property is mandatory.

> Issue 2: Property JNDI Context Factory is set to mandatory.

> Issue 3: JMS ObjectMessage is indicated in the PRD as one of the message types to be supported by this plugin. As far as I researched and tested, ObjectMessage usage is generally discouraged, as it introduces coupling of class paths between producers and consumers. This implies having the same class paths in your producer and consumer (JMS plugin in CDAP) side. For more information: http://activemq.apache.org/objectmessage.html. 

> Issue 4: Currently, monitoring is integrated only to few of Google plugins (as for example BigQuery). All the plugins with monitoring included are batch plugins (no real-time plugins with monitoring included so far).
When a batch plugin finishes running, it executes onRunFinish()method which counts the number of records that are read from the source (or written to a sink) and writes a metric in SparkContext. 
In the streaming plugins there are some points of concern:
JMS does not provide a built-in functionality to count the total number of messages residing in a topic/queue.
A streaming plugin by default runs forever. It only gets stopped when the user stops the execution manually. In such cases, onRunFinish() method is not called (as far as we have tested).
One solution we tried was to increase a metric for delta=1, every time a message is processed correctly. This required us to pass a StreamingContext message to the JMSReceiver class, which threw an NotSerializableException.

> Issue 5: When consuming a message, should we consume only the payload or also the headers? Headers are automatically generated when a JMS message is produced.

> Issue 6: No logos received for the plugin.

3. Removed schema on the sink side. Since I maintain a single config class for both  source/sink plugins, the schema was included also at the sink plugin 😬.

4. Fixed this. Reference name now is at the top.

CuriousVini commented on src/main/java/io/cdap/plugin/jms/common/JMSConfig.java:
```diff
@@ -0,0 +1,248 @@
+    this.jndiUsername = jndiUsername;
+    this.jndiPassword = jndiPassword;
+    this.messageType = Strings.isNullOrEmpty(messageType) ? JMSMessageType.TEXT.getName() : messageType;
+    this.destination = destination;
```
destination is assigned to itself?

CuriousVini commented on docs/JMS-batchsink.md:
```diff
@@ -0,0 +1,58 @@
+-------
+This example reads a record from a JSON file and produces a *MapMessage* with the file's content as a payload.
+For every record of the JSON file it produces one *MapMessage* to the given topic.
+An example of a JSON record is shown below.
```
Is this documentation valid for batch sink?

dardanxh commented:
@CuriousVini I've done some refactoring in the code and included those new requirements:

- Added a toggle widget to skip/keep header data of JMS messages.

- Added the functionality to skip null fields automatically (ie., will not get forwarded in the pipeline)

- Added support for Object messages. They are considered as byte arrays.

- Added the functionality to support durable subscribers. A durable subscriber only supports clients consuming messages from topics. Durable subscribers are not available for queues. If a consumer is a durable subscriber, it can get shut down, and still continue consuming message from where it left when it starts up. A consumer subscriber causes the broker (ie., ActiveMQ) to keep the state of messages consumed. More [here](https://docs.oracle.com/cd/E19798-01/821-1841/bncgd/index.html#:~:text=A%20durable%20subscriber%20registers%20a,retained%20by%20the%20JMS%20provider.&text=If%20a%20durable%20subscription%20has,subscription%20or%20until%20they%20expire.).

- JNDI Context Factory has `org.apache.activemq.jndi.ActiveMQInitialContextFactory` set as default value.

- Provide URL has `tcp://hostname:61616` set as a placeholder.

- Still doing some research on how to consume an entire history of messages from a queue/topic. For the queue case, I've already got on small example which I need to integrate.

CuriousVini commented:
Could you please update the design doc? 
Its hard to review ~3k lines of code without understanding the design.

dardanxh commented:
@CuriousVini I've done some refactoring on the sink side. Could you check.
